docs/slide-in-use-cases.md

```diff
@@ -0,0 +1,64 @@
+# Overview through Slide In use cases
+
+It turned out that the sliding-in feature is more complex than it had appeared in the first place.
+This paper is an overview through the use cases.
+It should help to implement and understand the feature.
+
+## Slide In New Branch Dialog
+
+The picture below shows the **Slide In New Branch Dialog**.
+This is a regular **Create New Branch Dialog** with a changed title.
+
+![Slide In New Branch Dialog](slide-in-new-branch-dialog.png "Slide In New Branch Dialog")
+
+Note the checkbox options:
+- Checkout branch
+- Overwrite existing branch
+
+## Use cases
+
+| UC ID | Branch entry exists | Local branch exists | Estimated frequency |
+| :---: | :---: | :---: | :---: |
+| 1 | false | false | ~ 75% |
+| 2 | false | true | ~ 25% |
+| 3 | true | true | < 1% |
+| 4 | true | false | < 1% |
+
+## Slide In - Logic
+
+- When the **entry does not exist**, it must be added (slid in) under the given branch.
+- When the **entry exists under the given parent**, branch layout must not be affected.
+- When the **entry exists under other parent**, it must be first removed (slid out),
+and then added (slid in) under the given parent.
+
+## Slide In - New Branches
+
+This section applies to **UC1** and **UC4**.
+
+
+Use case **UC1** seems to be the most common.
+They are simply added in the way specified by the [slide in logic](#slide-in---logic).
+
+
+
+## Slide In - With "Overwrite existing branch"
+
+This section applies to **UC2** and **UC3**.
+
+
+**Overwriting a parent branch must not affect the branch layout**.
+
+
+The other cases are described below.
+
+| Branch State | Checkout | Branch Layout |
+| --- | --- | --- |
+| **Working tree changed** | no checkout | branch entry **added*** under given parent |
+| Branch contains **some commits** | no checkout | branch entry **added*** under given parent |
```

Comment:
`Branch contains **some commits**`... as in, branch tip is not equal to its fork point? :thinking:

Reply:
Does this row specifically refer to the case when this kind of message appears?

![image](https://user-images.githubusercontent.com/3383210/87127027-82de3780-c28d-11ea-846a-aa548504b746.png)

Reply:
Although I have no idea based on what criteria git4idea decided that "some commits can be lost"

Reply:
Steps to reproduce the above:
1. set up machete-sandbox from git-machete-blog-2
2. add `foo123` under `call-ws`
3. try sliding in under `hotfix/add-trigger`

Reply:
(Note that as a last resort there's also a nuke-it-all solution: forbid reattaching existing branches completely and just limit ourselves to adding new branches)

Reply:
> Does this row specifically refer to the case when this kind of message appears?
> 
> ![image](https://user-images.githubusercontent.com/3383210/87127027-82de3780-c28d-11ea-846a-aa548504b746.png)

yup, this is the expected behavior 

This (and "Working tree changed" aka uncommitted changes) may cause failed checkout notification (but successfully slide in an existing branch entry!).

Reply:
My (probably wrong) thinking was that we can cover all cases 
\+ more by enhancing the existing new branch dialog by performing our slide in related operations (consistency, the same checkboxes options provided)
\- than creating a custom slide in dialog (unachievable consistency, more work to provide the checkbox options)

I would like to rethink things (on Monday) and consider **nuke-it-all solution**.

Reply:
One more thing about the present solution cause it may not have been clear. 

Sliding in of the existing branch is now possible only with **overwrite** option (without it the `Create` button is disabled for existing branches).

Reply:
Thinking out loud.

We could create a `Slide In Dialog` that takes the name (any valid) and slides in under the specified parent.
Then, for the NON-EXISTING BRANCHES ONLY show the `Create New Branch Dialog` using the provided name as the initial field value.
- a user may change the name within the second dialog (it would be rather hard to forbid the edition)
- we omit the problematic overwriting case
- TBD what to do in case of second dialog cancelation: do we want the whole action to be atomic or not